src/resolve/target.rs

```diff
@@ -0,0 +1,331 @@
+                    .error("couldn't resolve file")
+                    .with_error_section(range, "defined here")
+                    .note(format!("tried to resolve {}", url))
+                    .note(format!("malformed reference: {}", err))
+                    .emit();
+                return Err(Error::Diagnostic);
+            },
+        };
+        let inner = match url.scheme() {
+            "heradoc" => match url.domain() {
+                Some("document") => TargetInner::LocalRelative(url.path_segments().unwrap().collect()),
+                Some(domain) => TargetInner::Implementation(domain.to_string()),
+                None => {
+                    diagnostics
+                        .error("no heradoc implementation domain found")
+                        .with_error_section(range, "defined here")
+                        .emit();
+                    return Err(Error::Diagnostic);
+                }
+            },
+            "file" => match url.to_file_path() {
+                Ok(path) => TargetInner::LocalAbsolute(path),
+                Err(()) => {
+                    diagnostics.error("error converting url to path")
+                        .with_info_section(range, "defined here")
+                        .error("the file url can't be converted to a path")
+                        .help("this could be due to a malformed URL like a non-empty or non-localhost domain")
+                        .emit();
+                    return Err(Error::Diagnostic);
+                }
+            },
+            _ => TargetInner::Remote(url.clone()),
+        };
+        Ok(Target {
+            inner,
+            meta: Meta {
+                url,
+                context,
+                project_root,
+                permissions,
+                range,
+                diagnostics,
+            }
+        })
+    }
+
+    pub fn canonicalize(self) -> Result<TargetCanonicalized<'a, 'd>> {
+        let Target { inner, meta } = self;
+
+        let canonicalize = |path: PathBuf| {
+            match path.canonicalize() {
+                Ok(path) => Ok(path),
+                Err(e) => {
+                    meta.diagnostics
+                        .error("error canonicalizing path")
+                        .with_error_section(meta.range, "trying to include this")
+                        .note(format!("canonicalizing the path: {:?}", path))
+                        .error(e.to_string())
+                        .emit();
+                    Err(Error::Diagnostic)
+                }
+            }
+        };
+
+        let inner = match inner {
+            inner @ TargetInner::Implementation(_) => inner,
+            inner @ TargetInner::Remote(_) => inner,
+            TargetInner::LocalAbsolute(abs) => TargetInner::LocalAbsolute(canonicalize(abs)?),
+            TargetInner::LocalRelative(rel) => {
+                assert!(rel.is_relative(), "TargetInner::LocalRelative not relative before canonicalizing: {:?}", rel);
```

Comment:
Should this not be a `bug` diagnostics, depending on how sure we are never to change this property?

Reply:
If we change this property, we need to change the lines below as well. This assertion ensures that such a change will be caught during testing. This guarantee is also written in the documentation of `LocalRelative`.

Reply:
You are probably right, this should be caught in the resolver checks if it were to change. Thus assertion instead of `bug!()` termination seems fine.